src/Targeting/ActionHandler/AssignTargetGroupAndSegment.php

```diff
@@ -95,9 +95,14 @@ public function apply(VisitorInfo $visitorInfo, array $action, Rule $rule = null
         }
 
         $targetGroupId = $action['targetGroup'] ?? null;
-        $targetGroup = TargetGroup::getById($targetGroupId);
+        if ($targetGroupId instanceof TargetGroup) {
+            $targetGroup = $action['targetGroup'] ?? null;
```

Comment:
```suggestion
            $targetGroup = $targetGroupId;
```
we don't need `??` here anymore, as we do that already a few lines above.